library/kotlin/src/io/envoyproxy/envoymobile/EnvoyBuilder.kt

```diff
@@ -0,0 +1,115 @@
+
+open class EnvoyBuilder internal constructor(
+    envoyConfiguration: EnvoyConfiguration
+) {
+  private var logLevel = LogLevel.INFO
+  private var template: String
+  private var engineType: () -> EnvoyEngine = { EnvoyEngineImpl() }
+
+  private var connectTimeoutSeconds = 30
+  private var dnsRefreshSeconds = 60
+  private var statsFlushSeconds = 60
+
+  constructor() : this(EnvoyConfigurationImpl())
+
+  init {
+    template = envoyConfiguration.templateString()
+  }
+
+  /**
+   * Add a log level to use with Envoy.
+   * @param logLevel the log level to use with Envoy.
+   */
+  fun addLogLevel(logLevel: LogLevel): EnvoyBuilder {
+    this.logLevel = logLevel
+    return this
+  }
+
+  /**
+   * Add a YAML file to use as a configuration.
+   * Setting this will supersede any other configuration settings in the builder.
+   *
+   * @param configYAML the YAML file to use as a configuration.
+   */
+  fun addConfigYAML(configYAML: String): EnvoyBuilder {
+    this.template = configYAML
+    return this
+  }
+
+  /**
+   * Add a timeout for new network connections to hosts in the cluster.
+   *
+   * @param connectTimeoutSeconds timeout for new network connections to hosts in the cluster.
+   */
+  fun withConnectTimeoutSeconds(connectTimeoutSeconds: Int): EnvoyBuilder {
+    this.connectTimeoutSeconds = connectTimeoutSeconds
+    return this
+  }
+
+  /**
+   * Add a rate at which to refresh DNS.
+   *
+   * @param dnsRefreshSeconds rate in seconds to refresh DNS.
+   */
+  fun withDNSRefreshSeconds(dnsRefreshSeconds: Int): EnvoyBuilder {
+    this.dnsRefreshSeconds = dnsRefreshSeconds
+    return this
+  }
+
+  /**
+   * Add an interval at which to flush Envoy stats.
+   *
+   * @param statsFlushSeconds interval at which to flush Envoy stats.
+   */
+  fun withStatsFlushSeconds(statsFlushSeconds: Int): EnvoyBuilder {
+    this.statsFlushSeconds = statsFlushSeconds
+    return this
+  }
+
+  /**
+   * Builds a new instance of Envoy using the provided configurations.
+   *
+   * @return A new instance of Envoy.
+   */
+  fun build(): Envoy {
+    return Envoy(engineType(), resolvedYAML(), logLevel)
+  }
+
+  /**
+   * Add a specific implementation of `EnvoyEngine` to use for starting Envoy.
+   *
+   * A new instance of this engine will be created when `build()` is called.
+   */
+  internal fun addEngineType(engineType: () -> EnvoyEngine): EnvoyBuilder {
+    this.engineType = engineType
+    return this
+  }
+
+
+  /** Processes the YAML template provided, replacing keys with values from the configuration.
+   *
+   * @parameter template: The template YAML file to use.
+   * @returns: A resolved YAML file with all template keys replaced.
+   * @throws ConfigurationException when the yaml configuration replacement is incomplete
+   */
+  private fun resolvedYAML(): String {
+    val resolvedTemplate = template
+        .replace("connect_timeout", connectTimeoutSeconds.toString())
```

Comment:
Templates are in the format of `{{ connect_timeout }}`, so I don't think this works?

Reply:
(The braces need to be removed because as-is you're replacing the keys as well as the values)

Reply:
Yeah, tests failed